File: python/signalhouse/domains/campaigns.py
"""Campaigns domain for the SignalHouse SDK."""

from __future__ import annotations

from typing import Any, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from ..client import SignalHouseSDK


class CampaignsAdmin:
    """Admin-only campaign operations (SignalHouse staff)."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk

    def approve_campaign(
        self,
        campaign_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Approve a campaign that is pending approval.

        Args:
            campaign_id: The ID of the campaign to approve.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_id is missing.
        """
        self._sdk._require({"campaignId": campaign_id})
        safe_campaign_id = quote(str(campaign_id), safe="")
        return self._sdk._request(
            f"/campaign/approve/{safe_campaign_id}",
            method="POST",
            token=token,
            headers=headers,
        )

    def reject_campaign(
        self,
        campaign_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Reject a campaign that is pending approval.

        Args:
            campaign_id: The ID of the campaign to reject.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_id is missing.
        """
        self._sdk._require({"campaignId": campaign_id})
        safe_campaign_id = quote(str(campaign_id), safe="")
        return self._sdk._request(
            f"/campaign/reject/{safe_campaign_id}",
            method="POST",
            token=token,
            headers=headers,
        )


class Campaigns:
    """10DLC Campaign management operations."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk
        self.admin: CampaignsAdmin | None = None
        if sdk.enable_admin:
            self.admin = CampaignsAdmin(sdk)

    def get_campaigns(
        self,
        *,
        id: str | None = None,
        brand_id: str | None = None,
        subgroup_id: str | None = None,
        group_id: str | None = None,
        page: int | None = None,
        limit: int | None = None,
        status: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get a list of campaigns with optional filters.

        Args:
            id: The ID of the campaign to filter by.
            brand_id: The ID of the brand to filter by.
            subgroup_id: The ID of the subgroup to filter by.
            group_id: The ID of the group to filter by.
            page: The page number for pagination.
            limit: The number of items per page.
            status: The status of the campaign to filter by.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({
            "id": id,
            "brandId": brand_id,
            "subgroupId": subgroup_id,
            "groupId": group_id,
            "page": page,
            "limit": limit,
            "status": status,
        })
        return self._sdk._request(
            f"/campaign{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def create_campaign(
        self,
        campaign_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Create a new campaign.

        Args:
            campaign_data: The data for the campaign to be created. Required fields include
                          brandId, usecase, description, messageFlow, privacyPolicyLink,
                          termsAndConditionsLink, phoneNumbers.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_data is missing.
        """
        self._sdk._require({"campaignData": campaign_data})
        return self._sdk._request(
            "/campaign",
            method="POST",
            body=campaign_data,
            token=token,
            headers=headers,
        )

    def update_campaign(
        self,
        campaign_id: str,
        campaign_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Update an existing campaign.

        Args:
            campaign_id: The ID of the campaign to update.
            campaign_data: The data for the campaign to be updated.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_id is missing.
        """
        self._sdk._require({"campaignId": campaign_id})
        safe_campaign_id = quote(str(campaign_id), safe="")
        return self._sdk._request(
            f"/campaign/{safe_campaign_id}",
            method="PUT",
            body=campaign_data,
            token=token,
            headers=headers,
        )

    def delete_campaign(
        self,
        campaign_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Delete an existing campaign (mark it as EXPIRED). The campaign will still be retrievable.

        Args:
            campaign_id: The ID of the campaign to delete.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_id is missing.
        """
        self._sdk._require({"campaignId": campaign_id})
        safe_campaign_id = quote(str(campaign_id), safe="")
        return self._sdk._request(
            f"/campaign/{safe_campaign_id}",
            method="DELETE",
            token=token,
            headers=headers,
        )

    def appeal_dca_rejection(
        self,
        campaign_id: str,
        appeal_data: dict[str, Any] | None = None,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Appeal a DCA-rejected campaign.

        Args:
            campaign_id: The ID of the campaign to appeal.
            appeal_data: Optional request body (e.g., {"reason": "..."}).
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_id is missing.
        """
        self._sdk._require({"campaignId": campaign_id})
        safe_campaign_id = quote(str(campaign_id), safe="")
        return self._sdk._request(
            f"/campaign/appealDcaRejection/{safe_campaign_id}",
            method="POST",
            body=appeal_data,
            token=token,
            headers=headers,
        )

    def nudge_dca_for_campaign(
        self,
        campaign_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Nudge a connectivity partner to prioritize review of a campaign.

        The campaign must be in PENDING_DCA_APPROVAL status.

        Args:
            campaign_id: The ID of the campaign to nudge.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If campaign_id is missing.
        """
        self._sdk._require({"campaignId": campaign_id})
        safe_campaign_id = quote(str(campaign_id), safe="")
        return self._sdk._request(
            f"/campaign/nudge/{safe_campaign_id}",
            method="POST",
            token=token,
            headers=headers,
        )

File: php/src/Domains/Campaigns.php
<?php

namespace SignalHouse\SDK\Domains;

use SignalHouse\SDK\HttpClient;

class Campaigns
{
    private HttpClient $client;
    private bool $enableAdmin;
    public ?object $admin = null;

    public function __construct(HttpClient $client, bool $enableAdmin)
    {
        $this->client = $client;
        $this->enableAdmin = $enableAdmin;

        if ($enableAdmin) {
            $this->admin = new class($client) {
                private HttpClient $client;

                public function __construct(HttpClient $client)
                {
                    $this->client = $client;
                }

                /**
                 * Approve a campaign that is pending approval
                 */
                public function approveCampaign(string $campaignId, array $options = []): array
                {
                    $this->client->require(['campaignId' => $campaignId]);
                    $safeCampaignId = rawurlencode($campaignId);
                    return $this->client->request("/campaign/approve/{$safeCampaignId}", array_merge([
                        'method' => 'POST',
                    ], $options));
                }

                /**
                 * Reject a campaign that is pending approval
                 */
                public function rejectCampaign(string $campaignId, array $options = []): array
                {
                    $this->client->require(['campaignId' => $campaignId]);
                    $safeCampaignId = rawurlencode($campaignId);
                    return $this->client->request("/campaign/reject/{$safeCampaignId}", array_merge([
                        'method' => 'POST',
                    ], $options));
                }
            };
        }
    }

    /**
     * Get a list of campaigns with optional filters
     *
     * @param array $params Filter parameters (id, brandId, subgroupId, groupId, page, limit, status)
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function getCampaigns(array $params = [], array $options = []): array
    {
        $queryString = $this->client->getQueryString($params);
        return $this->client->request("/campaign{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Create a new campaign
     *
     * @param array $campaignData The campaign data (see JS SDK CreateCampaignData for fields)
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function createCampaign(array $campaignData, array $options = []): array
    {
        $this->client->require(['campaignData' => $campaignData]);
        return $this->client->request('/campaign', array_merge([
            'method' => 'POST',
            'body' => $campaignData,
        ], $options));
    }

    /**
     * Update an existing campaign
     *
     * @param string $campaignId The ID of the campaign
     * @param array $campaignData The data to update
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function updateCampaign(string $campaignId, array $campaignData, array $options = []): array
    {
        $this->client->require(['campaignId' => $campaignId]);
        $safeCampaignId = rawurlencode($campaignId);
        return $this->client->request("/campaign/{$safeCampaignId}", array_merge([
            'method' => 'PUT',
            'body' => $campaignData,
        ], $options));
    }

    /**
     * Delete a campaign (mark as EXPIRED)
     *
     * @param string $campaignId The ID of the campaign
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function deleteCampaign(string $campaignId, array $options = []): array
    {
        $this->client->require(['campaignId' => $campaignId]);
        $safeCampaignId = rawurlencode($campaignId);
        return $this->client->request("/campaign/{$safeCampaignId}", array_merge([
            'method' => 'DELETE',
        ], $options));
    }

    /**
     * Appeal a DCA rejection for a campaign.
     *
     * @param string $campaignId The ID of the campaign
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function appealDcaRejection(string $campaignId, array $options = []): array
    {
        $this->client->require(['campaignId' => $campaignId]);
        $safeCampaignId = rawurlencode($campaignId);
        return $this->client->request("/campaign/appealDcaRejection/{$safeCampaignId}", array_merge([
            'method' => 'POST',
        ], $options));
    }

    /**
     * Nudge a connectivity partner to prioritize review of a campaign.
     *
     * @param string $campaignId The ID of the campaign
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function nudgeDcaForCampaign(string $campaignId, array $options = []): array
    {
        $this->client->require(['campaignId' => $campaignId]);
        $safeCampaignId = rawurlencode($campaignId);
        return $this->client->request("/campaign/nudge/{$safeCampaignId}", array_merge([
            'method' => 'POST',
        ], $options));
    }
}
